File: crates/tcode-tui/src/usage.rs
//! Token accounting for the status row and the turn receipt.
//!
//! The two figures here are different quantities and must not be mixed (see
//! `plan.md`): `context` is what a single request's full prompt occupies in
//! the model window — cached prefix included, because cached input still
//! takes up room. `turn` is the receipt for one turn, and reports the
//! *uncached* input actually paid for plus a cache-hit share. Summing
//! `total_input()` across a multi-step turn would recount the cached prefix
//! once per request.

use std::time::{Duration, Instant, SystemTime, UNIX_EPOCH};

use ratatui::text::{Line, Span};
use tcode_core::{RateLimits, Usage};

use crate::theme;

/// Everything the UI knows about what the running turn has cost.
///
/// These fields move together — a turn reset touches all of them, a usage
/// event replaces several at once — so they live behind one type rather than
/// as ten loose `App` fields that any method could half-update.
pub struct TurnMeter {
    /// Streamed output tokens, for the live `↓ ~N tok` readout.
    pub out_tokens: usize,
    /// Usage reported by sub-agents. Tracked apart from `turn` so it can be
    /// re-added after the session's authoritative tally replaces the estimate.
    delegated: Usage,
    pub turn: Usage,
    pub rate_limits: Option<RateLimits>,
    /// Best available estimate of the conversation currently occupying the
    /// model window. A completed provider usage event replaces estimates;
    /// streamed output and tool results keep it moving between those events.
    pub context_tokens: u64,
    /// Start of the current model request, so a retry can discard the
    /// speculative streamed-token estimate from the failed attempt.
    step_start: u64,
    /// Session JSONL stores messages, not provider token counters. A resumed
    /// conversation starts from a local estimate until its next response
    /// supplies an authoritative usage event.
    pub context_estimated: bool,
    /// Time the turn was deliberately paused for a human decision. Receipts
    /// report active execution time, not time spent deciding.
    wait_started: Option<Instant>,
    wait_total: Duration,
    /// Cache-read share of the previous turn; the regression sentinel compares
    /// against it so cache decay is visible immediately.
    prev_cache_ratio: Option<f64>,
}

impl TurnMeter {
    pub fn new(context_tokens: u64, context_estimated: bool) -> Self {
        Self {
            out_tokens: 0,
            delegated: Usage::default(),
            turn: Usage::default(),
            rate_limits: None,
            context_tokens,
            step_start: context_tokens,
            context_estimated,
            wait_started: None,
            wait_total: Duration::ZERO,
            prev_cache_ratio: None,
        }
    }

    /// Zero the per-turn tallies. Context figures survive: they describe the
    /// conversation, not the turn.
    pub fn start_turn(&mut self) {
        self.turn = Usage::default();
        self.delegated = Usage::default();
        self.wait_started = None;
        self.wait_total = Duration::ZERO;
        self.out_tokens = 0;
    }

    /// Compaction legitimately rewrites the prefix, so the next turn's low
    /// cache share is expected rather than a regression to warn about.
    pub fn forget_cache_baseline(&mut self) {
        self.prev_cache_ratio = None;
    }

    /// Streamed text, thinking or tool-input deltas: they are in the window
    /// before any usage event confirms them.
    pub fn on_streamed_tokens(&mut self, tokens: usize) {
        self.out_tokens += tokens;
        self.context_tokens = self.context_tokens.saturating_add(tokens as u64);
    }

    pub fn add_context(&mut self, tokens: u64) {
        self.context_tokens = self.context_tokens.saturating_add(tokens);
    }

    /// An authoritative provider tally: it replaces the speculative estimate
    /// rather than adding to it.
    pub fn on_usage(&mut self, u: Usage) {
        self.turn.input_tokens += u.input_tokens;
        self.turn.output_tokens += u.output_tokens;
        self.turn.cache_read_tokens += u.cache_read_tokens;
        self.turn.cache_write_tokens += u.cache_write_tokens;
        self.context_tokens = u.total_input().saturating_add(u.output_tokens);
        self.step_start = self.context_tokens;
        self.context_estimated = false;
    }

    pub fn on_delegated_usage(&mut self, u: Usage) {
        self.delegated = add_usage(self.delegated, u);
        self.turn = add_usage(self.turn, u);
        self.out_tokens = self.out_tokens.saturating_add(u.output_tokens as usize);
    }

    /// A new request begins: the current estimate becomes the point a retry
    /// rewinds to.
    pub fn begin_step(&mut self) {
        self.step_start = self.context_tokens;
    }

    /// A failed attempt streamed tokens that never became part of a prompt.
    pub fn rewind_step(&mut self) {
        self.context_tokens = self.step_start;
    }

    pub fn set_context(&mut self, tokens: u64, estimated: bool) {
        self.context_tokens = tokens;
        self.step_start = tokens;
        self.context_estimated = estimated;
    }

    pub fn pause_for_user(&mut self) {
        if self.wait_started.is_none() {
            self.wait_started = Some(Instant::now());
        }
    }

    pub fn resume_from_user(&mut self) {
        if let Some(started) = self.wait_started.take() {
            self.wait_total += started.elapsed();
        }
    }

    /// Wall time minus every stretch spent waiting on a human decision.
    pub fn active_elapsed(&self, started: Instant) -> f32 {
        started
            .elapsed()
            .saturating_sub(self.wait_total)
            .saturating_sub(
                self.wait_started
                    .map(|wait| wait.elapsed())
                    .unwrap_or_default(),
            )
            .as_secs_f32()
    }

    /// Adopt the session's authoritative per-turn tally, which also covers
    /// compaction (that streams no Usage events to the UI).
    pub fn finish_turn(&mut self, session_usage: Usage) {
        self.turn = add_usage(session_usage, self.delegated);
    }

    /// Cache regression sentinel: an append-only ledger should keep the hit
    /// share high, so a sharp drop means something rewrote the prefix and
    /// deserves attention now, not on the monthly bill. Returns the previous
    /// and current share when the drop is worth reporting.
    pub fn take_cache_regression(&mut self) -> Option<(f64, f64)> {
        if self.turn.total_input() == 0 {
            return None;
        }
        let ratio = self.turn.cache_read_tokens as f64 / self.turn.total_input() as f64;
        let regression = self
            .prev_cache_ratio
            .filter(|prev| *prev >= 0.5 && ratio < prev * 0.5)
            .map(|prev| (prev, ratio));
        self.prev_cache_ratio = Some(ratio);
        regression
    }
}

/// One compact row below the editor. The meter intentionally reports the
/// current conversation, rather than cumulative billable tokens: cached input
/// still occupies context and must count toward the model window.
pub fn context_progress_line(
    used: u64,
    window: u64,
    terminal_width: u16,
    estimated: bool,
) -> Line<'static> {
    let window = window.max(1);
    let pct = used.saturating_mul(100).saturating_div(window).min(100);
    let estimate_mark = if estimated { "≈" } else { "" };
    let color = if pct >= 95 {
        theme::ERROR
    } else if pct >= 85 {
        theme::WARN
    } else {
        theme::OK
    };
    let (label, bar_width) = if terminal_width < 42 {
        ("  ctx ", 8usize)
    } else {
        ("  context ", 12usize)
    };
    let filled = if used == 0 {
        0
    } else {
        ((bar_width as u64 * pct).div_ceil(100) as usize).min(bar_width)
    };
    let mut spans = vec![Span::styled(label, theme::dim())];
    spans.extend(slim_bar(filled, bar_width, color));
    spans.push(Span::styled(
        format!(" {estimate_mark}{pct}%"),
        ratatui::style::Style::default().fg(color),
    ));
    if terminal_width >= 42 {
        spans.push(Span::styled(
            format!(" · {}", token_count(window)),
            theme::dim(),
        ));
    }
    Line::from(spans)
}

/// The slim gauge shared by the context meter and the rate-limit row:
/// a heavy coloured run over a dim dashed track.
fn slim_bar(filled: usize, width: usize, color: ratatui::style::Color) -> [Span<'static>; 2] {
    [
        Span::styled(
            "━".repeat(filled),
            ratatui::style::Style::default().fg(color),
        ),
        Span::styled("╌".repeat(width.saturating_sub(filled)), theme::dim()),
    ]
}

pub fn token_count(tokens: u64) -> String {
    if tokens < 1_000 {
        tokens.to_string()
    } else if tokens < 10_000 {
        format!("{:.1}k", tokens as f64 / 1_000.0)
    } else {
        format!("{}k", tokens.div_ceil(1_000))
    }
}

pub fn rate_limit_line(limits: RateLimits) -> Line<'static> {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |duration| duration.as_secs());
    rate_limit_line_at(limits, now)
}

fn rate_limit_line_at(limits: RateLimits, now: u64) -> Line<'static> {
    let primary_used = limits.primary.used_percent.clamp(0.0, 100.0);
    let filled = ((primary_used / 100.0) * 12.0).round() as usize;
    let color = usage_color(primary_used);
    let mut spans = vec![Span::styled("  Codex 5h ", theme::dim())];
    spans.extend(slim_bar(filled, 12, color));
    spans.push(Span::styled(
        format!(" {primary_used:.0}%"),
        ratatui::style::Style::default().fg(color),
    ));
    append_reset_countdown(&mut spans, limits.primary.resets_at, now);

    if let Some(weekly) = limits.secondary.filter(|limit| limit.used_percent >= 80.0) {
        let weekly_used = weekly.used_percent.clamp(0.0, 100.0);
        let weekly_filled = ((weekly_used / 100.0) * 12.0).round() as usize;
        let weekly_color = usage_color(weekly_used);
        spans.push(Span::styled(" · week ", theme::dim()));
        spans.extend(slim_bar(weekly_filled, 12, weekly_color));
        spans.push(Span::styled(
            format!(" {weekly_used:.0}%"),
            ratatui::style::Style::default().fg(weekly_color),
        ));
        append_reset_countdown(&mut spans, weekly.resets_at, now);
    }
    Line::from(spans)
}

fn usage_color(used_percent: f64) -> ratatui::style::Color {
    if used_percent >= 90.0 {
        theme::ERROR
    } else if used_percent >= 75.0 {
        theme::WARN
    } else {
        theme::OK
    }
}

fn append_reset_countdown(spans: &mut Vec<Span<'static>>, resets_at: u64, now: u64) {
    let Some(remaining) = resets_at.checked_sub(now).filter(|&seconds| seconds > 0) else {
        return;
    };
    spans.push(Span::styled(
        format!(" ↻ {}", brief_duration(remaining)),
        theme::dim(),
    ));
}

/// Compact countdown for the status line: enough precision for a human to
/// decide whether to wait, without turning the meter into a timestamp.
fn brief_duration(seconds: u64) -> String {
    if seconds < 60 {
        "<1m".into()
    } else if seconds < 3_600 {
        format!("{}m", seconds.div_ceil(60))
    } else if seconds < 86_400 {
        format!("{}h{}m", seconds / 3_600, (seconds % 3_600).div_ceil(60))
    } else {
        format!("{}d", seconds.div_ceil(86_400))
    }
}

pub fn add_usage(left: Usage, right: Usage) -> Usage {
    Usage {
        input_tokens: left.input_tokens.saturating_add(right.input_tokens),
        output_tokens: left.output_tokens.saturating_add(right.output_tokens),
        cache_read_tokens: left
            .cache_read_tokens
            .saturating_add(right.cache_read_tokens),
        cache_write_tokens: left
            .cache_write_tokens
            .saturating_add(right.cache_write_tokens),
    }
}

pub fn turn_summary_line(elapsed: f32, usage: Usage) -> Line<'static> {
    let cache_pct = if usage.total_input() > 0 {
        (usage.cache_read_tokens as f64 / usage.total_input() as f64 * 100.0).round()
    } else {
        0.0
    };
    let cache_style = if cache_pct > 0.0 {
        theme::accent()
    } else {
        theme::dim()
    };
    Line::from(vec![
        Span::styled("✓ completed ", theme::ok()),
        Span::styled(format!("{elapsed:.1}s"), theme::bold()),
        Span::styled(" · ↑", theme::dim()),
        // Uncached input only: the tokens this turn actually paid full price
        // for. Summing total_input() across a multi-step turn would recount
        // the cached prefix on every request; the cache figure below shows
        // how much of the full prompt was reused. This is a turn receipt, not
        // the window-occupancy figure the context meter reports.
        Span::styled(token_count(usage.input_tokens), theme::accent()),
        Span::styled(" · ↓", theme::dim()),
        Span::styled(
            token_count(usage.output_tokens),
            ratatui::style::Style::default().fg(theme::OK),
        ),
        Span::styled(" · cache ", theme::dim()),
        Span::styled(format!("{cache_pct:.0}%"), cache_style),
    ])
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn normal_usage_meters_use_the_tool_green() {
        let context = context_progress_line(20_000, 200_000, 80, false);
        assert!(context
            .spans
            .iter()
            .any(|span| span.style.fg == Some(theme::OK)));

        let limits = tcode_core::RateLimits {
            primary: tcode_core::RateLimit {
                used_percent: 30.0,
                window_minutes: 300,
                resets_at: 14_800,
            },
            secondary: None,
        };
        let rate_limit = rate_limit_line_at(limits, 10_000);
        assert!(rate_limit
            .spans
            .iter()
            .any(|span| span.style.fg == Some(theme::OK)));
    }

    #[test]
    fn context_meter_reports_percent_and_warning_color() {
        let line = context_progress_line(170_000, 200_000, 80, false);
        let text = line
            .spans
            .iter()
            .map(|span| span.content.as_ref())
            .collect::<String>();
        assert!(text.contains("context"));
        assert!(text.contains("85% · 200k"));
        assert!(!text.contains("170k/200k"));
        assert!(line
            .spans
            .iter()
            .any(|span| span.style.fg == Some(theme::WARN)));
    }

    #[test]
    fn codex_rate_limit_line_shows_used_percent_and_reset_countdowns() {
        let limits = tcode_core::RateLimits {
            primary: tcode_core::RateLimit {
                used_percent: 30.0,
                window_minutes: 300,
                resets_at: 14_800,
            },
            secondary: Some(tcode_core::RateLimit {
                used_percent: 80.0,
                window_minutes: 10_080,
                resets_at: 269_200,
            }),
        };
        let text = rate_limit_line_at(limits, 10_000)
            .spans
            .iter()
            .map(|span| span.content.as_ref())
            .collect::<String>();

        assert!(text.contains("Codex 5h"));
        assert!(text.contains(" 30% ↻ 1h20m"));
        assert!(text.contains("week "));
        assert!(text.contains(" 80% ↻ 3d"));
    }

    #[test]
    fn brief_duration_stays_compact_at_unit_boundaries() {
        assert_eq!(brief_duration(59), "<1m");
        assert_eq!(brief_duration(60), "1m");
        assert_eq!(brief_duration(3_601), "1h1m");
        assert_eq!(brief_duration(86_401), "2d");
    }

    #[test]
    fn codex_rate_limit_line_hides_week_below_80_percent() {
        let limits = tcode_core::RateLimits {
            primary: tcode_core::RateLimit {
                used_percent: 30.0,
                window_minutes: 300,
                resets_at: 0,
            },
            secondary: Some(tcode_core::RateLimit {
                used_percent: 79.9,
                window_minutes: 10_080,
                resets_at: 0,
            }),
        };
        let text = rate_limit_line(limits)
            .spans
            .iter()
            .map(|span| span.content.as_ref())
            .collect::<String>();

        assert!(!text.contains("week"));
    }

    #[test]
    fn turn_summary_is_a_scannable_receipt() {
        let line = turn_summary_line(
            2.5,
            Usage {
                input_tokens: 1_178,
                output_tokens: 23,
                ..Usage::default()
            },
        );
        let text = line
            .spans
            .iter()
            .map(|span| span.content.as_ref())
            .collect::<String>();
        assert_eq!(text, "✓ completed 2.5s · ↑1.2k · ↓23 · cache 0%");
    }

    #[test]
    fn delegated_usage_is_added_without_losing_cache_fields() {
        let total = add_usage(
            Usage {
                input_tokens: 10,
                output_tokens: 2,
                cache_read_tokens: 3,
                cache_write_tokens: 4,
            },
            Usage {
                input_tokens: 20,
                output_tokens: 5,
                cache_read_tokens: 6,
                cache_write_tokens: 7,
            },
        );
        assert_eq!(total.input_tokens, 30);
        assert_eq!(total.output_tokens, 7);
        assert_eq!(total.cache_read_tokens, 9);
        assert_eq!(total.cache_write_tokens, 11);
    }

    /// A retry must not leave the failed attempt's speculative streamed
    /// tokens counted against the window.
    #[test]
    fn a_rewound_step_discards_speculative_streamed_tokens() {
        let mut meter = TurnMeter::new(1_000, false);
        meter.begin_step();
        meter.on_streamed_tokens(250);
        assert_eq!(meter.context_tokens, 1_250);
        meter.rewind_step();
        assert_eq!(meter.context_tokens, 1_000);
    }

    /// The sentinel fires once per drop and then re-baselines, so a genuinely
    /// cold prefix does not warn on every following turn.
    #[test]
    fn cache_regression_reports_a_sharp_drop_then_rebaselines() {
        let mut meter = TurnMeter::new(0, false);
        meter.turn = Usage {
            input_tokens: 100,
            cache_read_tokens: 900,
            ..Usage::default()
        };
        assert!(
            meter.take_cache_regression().is_none(),
            "first turn sets a baseline"
        );

        meter.turn = Usage {
            input_tokens: 900,
            cache_read_tokens: 100,
            ..Usage::default()
        };
        let (prev, now) = meter.take_cache_regression().expect("drop is reported");
        assert!(prev > 0.5 && now < 0.5);

        // Same low share again: already the baseline, so no second warning.
        assert!(meter.take_cache_regression().is_none());
    }

    /// Receipts report execution time, not time the human spent deciding.
    #[test]
    fn time_spent_awaiting_a_decision_is_excluded_from_the_receipt() {
        let mut meter = TurnMeter::new(0, false);
        let started = Instant::now();
        meter.pause_for_user();
        std::thread::sleep(Duration::from_millis(30));
        meter.resume_from_user();
        assert!(meter.active_elapsed(started) < 0.03);
    }
}

File: crates/tcode-app/ui/src/usage.ts
/**
 * What a turn costs and what is left of the budget, as pure reads.
 *
 * Two different quantities live here and must not be mixed (the same split
 * `tcode-tui/src/usage.rs` documents): **context** is what the conversation
 * occupies in the model's window right now — cached prefix included, because
 * cached input still takes up room — and **turn** is the receipt for one turn,
 * which reports the uncached input actually paid for. Summing `total_input()`
 * across a multi-step turn would recount the cached prefix once per request.
 *
 * The reducer is here rather than in `App.tsx` for this codebase's usual
 * reason: "an authoritative usage event replaces the estimate instead of adding
 * to it" is a rule with a right and a wrong answer, and answers are testable.
 * `UsagePanel.tsx` draws what these return.
 */

import type { AgentEvent } from "./types";

/** One provider budget window, as `tcode_core::RateLimit` sends it. */
export type Limit = {
  used_percent: number;
  /** 0 when the provider did not say. */
  window_minutes: number;
  /** Unix seconds; 0 when the provider did not say. */
  resets_at: number;
};

export type Limits = { primary: Limit; secondary: Limit | null };

/** Normalized token counts for one model request. Mirrors `tcode_core::Usage`:
 *  `input_tokens` is the NON-cached input only. */
export type Usage = {
  input_tokens: number;
  output_tokens: number;
  cache_read_tokens: number;
  cache_write_tokens: number;
};

export type Meter = {
  /** Best available estimate of what this conversation occupies in the window. */
  context: number;
  /**
   * True while that figure has never been confirmed by a provider tally.
   *
   * A resumed conversation starts here: the session log stores messages, not
   * token counters, so until the next response arrives the meter is showing an
   * estimate and has to say so rather than quietly rounding it to a fact.
   */
  estimated: boolean;
  /** The turn in flight, or the last one that finished. */
  turn: Usage;
};

export const NO_USAGE: Usage = {
  input_tokens: 0,
  output_tokens: 0,
  cache_read_tokens: 0,
  cache_write_tokens: 0,
};

export const NO_METER: Meter = {
  context: 0,
  estimated: false,
  turn: NO_USAGE,
};

/**
 * The subscription budget an event reports, or `null` when it reports none.
 *
 * Deliberately *not* part of `Meter`: a 5-hour window belongs to the account,
 * not to a conversation, and every session in the window spends the same one.
 * Keeping it per session would mean a folder opened five minutes ago claims to
 * know nothing about a budget the pane next to it is already showing at 42%.
 */
export function limitsFrom(event: AgentEvent): Limits | null {
  return event.type === "RateLimits" ? (event.data as Limits) : null;
}

/**
 * Fold one event into the meter.
 *
 * `Usage` is a per-step tally and is authoritative: the whole prompt for that
 * request is `total_input`, so the context figure is *replaced* by it rather
 * than accumulated. The turn receipt is the one thing that adds up, and
 * `DelegatedUsage` (a sub-agent's spend) joins it — a sub-agent costs money but
 * occupies its own window, not the parent's.
 *
 * Nothing here zeroes the receipt. `AgentEvent::Started` looks like the moment
 * to do it and is not: core emits it for every model *request*, so a turn that
 * takes six steps would arrive with only the sixth one's cost. The receipt is
 * reset where the turn is submitted (`App.tsx`), which is also where the TUI
 * does it.
 */
export function applyUsage(meter: Meter, event: AgentEvent): Meter {
  switch (event.type) {
    case "Usage": {
      const usage = event.data as Usage;
      return {
        ...meter,
        context: totalInput(usage) + usage.output_tokens,
        estimated: false,
        turn: addUsage(meter.turn, usage),
      };
    }
    case "DelegatedUsage":
      return { ...meter, turn: addUsage(meter.turn, event.data as Usage) };
    // Reference context is expanded into the prompt *before* the turn runs, so
    // it is already in the window while the meter is still showing whatever the
    // last response reported. Counting it keeps the ring honest for the one
    // case where a single message moves the needle several percent.
    case "ReferencesExpanded": {
      const added = (event.data as { added_tokens: number }).added_tokens;
      return { ...meter, context: meter.context + added, estimated: true };
    }
    // A compaction replaced the history with a summary. Nothing here knows how
    // big that summary is; the next response does, and until then saying so is
    // better than showing the pre-compaction figure as if it still held.
    case "Compacted":
      return { ...meter, estimated: true };
    default:
      return meter;
  }
}

/**
 * A resumed conversation's window occupancy, from the log itself.
 *
 * A session log stores messages, not token counters, so there is no true answer
 * to give until the next response arrives — and a resumed conversation with
 * 90k of history showing `0%` would be the one moment the meter is read and the
 * one moment it is wrong. Same rough divisor as core's `approx_tokens`, and the
 * result is always flagged `estimated`: it under-counts (the harness prefix and
 * project instructions never reach this side) and the first real tally replaces
 * it.
 */
export function estimateContext(history: unknown[]): number {
  if (history.length === 0) return 0;
  return Math.ceil(JSON.stringify(history).length / 3);
}

export function totalInput(usage: Usage): number {
  return usage.input_tokens + usage.cache_read_tokens + usage.cache_write_tokens;
}

function addUsage(left: Usage, right: Usage): Usage {
  return {
    input_tokens: left.input_tokens + right.input_tokens,
    output_tokens: left.output_tokens + right.output_tokens,
    cache_read_tokens: left.cache_read_tokens + right.cache_read_tokens,
    cache_write_tokens: left.cache_write_tokens + right.cache_write_tokens,
  };
}

/** Share of the last turn's prompt that was served from cache. `null` when the
 *  turn read nothing — 0% and "no input yet" are different facts. */
export function cacheShare(usage: Usage): number | null {
  const total = totalInput(usage);
  return total > 0 ? usage.cache_read_tokens / total : null;
}

/** 0–100, clamped. A window of 0 (no model, or a model with none declared)
 *  reads as empty rather than dividing by zero. */
export function percent(used: number, of: number): number {
  if (of <= 0) return 0;
  return Math.min(100, Math.max(0, (used / of) * 100));
}

/**
 * How loud a meter is allowed to be.
 *
 * Achromatic until it matters, which is this app's palette rule rather than a
 * shade of the terminal's: chroma means state, and "34% of a window used" is not
 * a state — it is the number being fine. The thresholds are the TUI's, so the
 * two frontends warn at the same moment.
 */
export type Level = "calm" | "high" | "full";

export function contextLevel(pct: number): Level {
  if (pct >= 95) return "full";
  if (pct >= 85) return "high";
  return "calm";
}

export function limitLevel(pct: number): Level {
  if (pct >= 90) return "full";
  if (pct >= 75) return "high";
  return "calm";
}

/** Compact token counts, as the TUI writes them: `980`, `1.2k`, `68k`. */
export function tokens(count: number): string {
  if (count < 1_000) return String(Math.round(count));
  if (count < 10_000) return `${(count / 1_000).toFixed(1)}k`;
  return `${Math.ceil(count / 1_000)}k`;
}

/**
 * What to call a budget window, from the minutes the provider reported.
 *
 * Derived rather than hard-coded to "5h" and "weekly": those are today's Codex
 * plan, and a meter that labels a 3-hour window "5h" because the string was
 * baked in is worse than one that says "3h".
 */
export function windowLabel(minutes: number): string {
  if (minutes <= 0) return "window";
  if (minutes >= 10_080 && minutes % 10_080 === 0) {
    const weeks = minutes / 10_080;
    return weeks === 1 ? "weekly" : `${weeks} weeks`;
  }
  if (minutes >= 1_440 && minutes % 1_440 === 0) {
    const days = minutes / 1_440;
    return days === 1 ? "daily" : `${days}d`;
  }
  if (minutes >= 60 && minutes % 60 === 0) return `${minutes / 60}h`;
  return `${minutes}m`;
}

/**
 * How long until a window resets, or `null` when it already has (or the
 * provider never said). `null` is the signal to draw nothing: a countdown that
 * has run out is noise, and "resets in 0m" reads as a fault.
 */
export function resetIn(resetsAt: number, now: number): string | null {
  const left = resetsAt - now;
  if (resetsAt <= 0 || left <= 0) return null;
  if (left < 60) return "under a minute";
  if (left < 3_600) return `${Math.ceil(left / 60)}m`;
  if (left < 86_400) {
    const hours = Math.floor(left / 3_600);
    const minutes = Math.ceil((left % 3_600) / 60);
    return minutes > 0 ? `${hours}h ${minutes}m` : `${hours}h`;
  }
  return `${Math.ceil(left / 86_400)}d`;
}
